### DISCUSSION TITLE
Using both HS and FS on an STM32H7R/STM32H7S with internal HS PHY; is this possible, could I port over code form another H7, or would there be issues?

### DISCUSSION THREAD

**Discussion by EamonBrennanDRS** [hathach/tinyusb#2813] [Q&A] [answered]
The basic idea I have is that my board would have two main ports: a HS port in the front for main communication (CDC, possibly DFU as well) and a FS port in the back (DFU only) that would also support USBPD; the FS part of the port would only be used for DFU and it would otherwise be plugged into a power adapter. The issue I am currently running into is that the H7RS line is considered distinct from the remaining H7 line, and, as such, has a completely different set up. It has an internal HS PHY, unlike the other H7 chips, so the family.c and similar files would not directly work with the H7RS line. 
Is there an alternative system that could work (not USBX, there's 0 documentation for what I am supposed to do, and what documentation does exist is incorrect), or should I make a request for TinyUSB to support the H7RS line?

**Comment by HiFiPhile** [comment]
Currently TUSB doesn't support using both ports as device concurrently, you have to choose the port on init. 
For H7RS support I hope it won't be hard, haven't looked into the detail but it should be similar to U5.

**Reply by EamonBrennanDRS** [reply]
Thanks. I think the plan I will go with is to use Cube's built in USB_DEVICE for the FS port and TinyUSB for the HS port. The built in USB_DEVICE is basically fully coded for the user, but only allows 1 port with 1 option, so the FS port will be used as a DFU, while the HS port will be dual rolled with CDC and DFU. 
The next question I would have is, can I just init TUSB as both CDC and DFU, or should I init CDC only, then deinit and swap to DFU with a command? I'm new to TUSB.

**Reply by HiFiPhile** [reply, marked as answer]
You can do any combination as long as it doesn't exceed the device endpoint and FIFO limit.
You can check cdc_dual_ports and cdc_msc example to see how it's done.
https://github.com/hathach/tinyusb/blob/f4dd1764849d005a2e44d51f62428aeaf2513804/examples/device/cdc_dual_ports/src/usb_descriptors.c#L133
https://github.com/hathach/tinyusb/blob/f4dd1764849d005a2e44d51f62428aeaf2513804/examples/device/cdc_msc/src/usb_descriptors.c#L144

**Reply by EamonBrennanDRS** [reply]
Thanks, I will try this out. I will also raise an issue for H7RS support.

**Comment by Lurcy38** [comment]
Hello hifiphile,
I need to create a quick demo on H7RS. I saw the H7RS is integrated in version 0.18.0 (saw OPT_MCU_STM32H7RS in the code), but no BSP support for H7RS is present in this release and in the master too. Do you know if there is any board support in alpha I can use for my quick project?

**Reply by HiFiPhile** [reply]
Yep, you can check #2960

**Comment by Lurcy38** [comment]
many thanks HifiPhile,
 i have analyzed the porting layer and i don't understand this piece of code 
void EXTI8_IRQHandler(void) {
    __HAL_GPIO_EXTI_CLEAR_IT(GPIO_PIN_8);
    if (tcpp0203_obj.IsInitialized) {
      TU_ASSERT(TCPP0203_SetPowerMode(&tcpp0203_obj, TCPP0203_POWER_MODE_NORMAL) == TCPP0203_OK, );
      TU_ASSERT(TCPP0203_SetGateDriverProvider(&tcpp0203_obj, TCPP0203_GD_PROVIDER_SWITCH_CLOSED) == TCPP0203_OK, );
    }
}
why do you need a EXTI Pin8 interrupt for this ?

regards

**Comment by Lurcy38** [comment]
forget my question :), i just understood the role of the int TCPP0203
